### PR TITLE
in/out @param javadoc

### PR DESCRIPTION
**Pull request by sdkrystian**

### CODE REVIEW

File: include/mrdox/Metadata/Javadoc.hpp [hunk 1]
```diff
@@ -65,6 +65,18 @@ struct MRDOX_VISIBLE
         warning
     };
 
+    /** Parameter pass direction.
+    */
+    enum class ParamDirection : int
+    {
+        // KRYSTIAN NOTE: no idea why this is needed,
+        // does not seem to break when changed to 0
+        none = 1, // needed by bitstream
```
Comment (klemens-morgenstern) [inline include/mrdox/Metadata/Javadoc.hpp:74]: does that mean unspecified?

File: source/api/_XML/XMLWriter.cpp [hunk 2]
```diff
@@ -615,8 +615,25 @@ XMLWriter::
...
+    switch(param.direction)
+    {
+    case Javadoc::ParamDirection::in:
+        direction = "in";
+        break;
+    case Javadoc::ParamDirection::out:
+        direction = "out";
+        break;
+    case Javadoc::ParamDirection::inout:
+        direction = "in,out";
```
Comment (klemens-morgenstern) [inline source/api/_XML/XMLWriter.cpp:628]: I think `inout` is preferrable

File: source/api/_XML/XMLWriter.cpp [hunk 3]
```diff
@@ -615,8 +615,25 @@ XMLWriter::
...
+        break;
+    default:
+        break;
+    }
     tags_.open("param", {
-        { "name", param.name, ! param.name.empty() }});
+        { "name", param.name, ! param.name.empty() },
+        { "direction", direction, !! direction }
```
Comment (klemens-morgenstern) [inline source/api/_XML/XMLWriter.cpp:635]: I'd prefer `direction != direction::none`.

File: test-files/old-tests/param-direction.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+/** test brief
+ * @param x
+ * @param[in] y
+**/
+void g(int x, int y) { }
+
+/**
+ * @param x
+ * @param[in,out] y
```
Comment (klemens-morgenstern) [inline test-files/old-tests/param-direction.cpp:14]: Doesn't clang handle `@param[inout]` ?

**Review by klemens-morgenstern — CHANGES_REQUESTED** [review]

File: include/mrdox/Metadata/Javadoc.hpp (hunk 1 above)
Comment (vinniefalco) [reply include/mrdox/Metadata/Javadoc.hpp:74]: I think it does and I think we can use 0 here

File: test-files/old-tests/param-direction.xml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  </function>
+  <function name="i" id="RmOJ/T+Nr/W5+22oXdFiYeoIDoA=">
+    <file path="param-direction.cpp" line="23" class="def"/>
+    <param name="x" type="int"/>
+    <param name="y" type="int"/>
+    <param name="z" type="int"/>
+    <doc>
+      <brief>
+      </brief>
+      <param name="x" direction="out">
```
Comment (vinniefalco) [inline test-files/old-tests/param-direction.xml:30]: `class="out"`

File: test-files/old-tests/param-direction.cpp (hunk 4 above)
Comment (sdkrystian) [reply test-files/old-tests/param-direction.cpp:14]: @vinniefalco @klemens-morgenstern it does not

File: source/lib/api/AST/ParseJavadoc.cpp [hunk 6]
```diff
@@ -334,6 +334,21 @@ class JavadocVisitor
             param.name = ensureUTF8(C->getParamNameAsWritten().str());
         else
             param.name = "@anon";
```
Comment (vinniefalco) [inline source/lib/api/AST/ParseJavadoc.cpp:336]: "@anon" ?

File: test-files/old-tests/param-direction.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
+/**
+ * @param x0
```
Comment (vinniefalco) [inline test-files/old-tests/param-direction.cpp:2]: When there are tests for javadocs, they should be strictly javadoc tests (I think yours are) and they should have a parent directory in the chain which is strictly javadoc tests

File: source/lib/api/AST/ParseJavadoc.cpp (hunk 6 above)
Comment (sdkrystian) [reply source/lib/api/AST/ParseJavadoc.cpp:336]: It's used for `@param` directives without any parameter name provided... only place I could see it being useful is for variadic function a la `void f(...)`.

File: source/api/_XML/XMLWriter.cpp (hunk 3 above)
Comment (vinniefalco) [reply source/api/_XML/XMLWriter.cpp]: Same.. !! ... ugh

**Pull request merged**